digits/model/tasks/caffe_train.py

```diff
@@ -118,6 +119,70 @@ def before_run(self):
+        mean_image = None
+        with open(self.dataset.path(mean_file), 'rb') as f:
+            blob = caffe_pb2.BlobProto()
+            blob.MergeFromString(f.read())
+
+            if isinstance(self.dataset, dataset.ImageClassificationDatasetJob):
+                mean_image = np.reshape(blob.data,
+                                        (
+                                            self.dataset.image_dims[2],
+                                            self.dataset.image_dims[0],
+                                            self.dataset.image_dims[1],
+                                        )
+                                    )
+            elif isinstance(self.dataset, dataset.GenericImageDatasetJob):
+                task = self.dataset.analyze_db_tasks()[0]
+                mean_image = np.reshape(blob.data,
+                                        (
+                                            task.image_channels,
+                                            task.image_height,
+                                            task.image_width,
+                                        )
+                                    )
+
+            # Resize the mean image if crop_size exists
+            if mean_image is not None and self.crop_size:
+                mean_image = mean_image.astype('uint8')
+                mean_image = mean_image.transpose(1,2,0)
+
+                shape = list(mean_image.shape)
+                # imresize will not resize if the depth is anything
+                # other than 3 or 4.  If it's 1, imresize expects an
+                # array.
+                if (len(shape) == 2 or (len(shape) == 3 and (shape[2] == 3 or shape[2] == 4))):
+                    mean_image = scipy.misc.imresize(mean_image, (self.crop_size, self.crop_size))
+                else:
+                    mean_image = scipy.misc.imresize(mean_image[:,:,0],
+                                                     (self.crop_size, self.crop_size))
+                    mean_image = np.expand_dims(mean_image, axis=2)
+                mean_image = mean_image.transpose(2,0,1)
+                mean_image = mean_image.astype('float')
+
+        return mean_image
+
+    def get_mean_pixel(self, mean_file):
+        mean_image = self.get_mean_image(mean_file)
+
+        mean_pixel = None
+        if mean_image is not None:
+            mean_pixel = mean_image.mean(1).mean(1)
+
+        return mean_pixel
+
+    def set_mean_value(self, layer, mean_pixel):
+        # remove any values that may already be in the network
+        layer.transform_param.ClearField('mean_file')
+        layer.transform_param.ClearField('mean_value')
+        layer.transform_param.mean_value.extend(list(mean_pixel))
+
+    def set_mean_file(self, layer, mean_file):
+        # remove any values that may already be in the network
+        layer.transform_param.ClearField('mean_value')
+        layer.transform_param.mean_file = mean_file
```

Comment:
These two functions do the correct thing from a functional point of view (at least I think they do - haven't tested it), but I'd like to print some WARNINGs to the log if any of these `ClearField()` calls are actually overwriting any information.